Automattic/jetpack
# My Jetpack: Update checkout flow to connect "after" checkout vs before.

This PR updates the My Jetpack product checkout flow to connect Jetpack "after" checkout is completed as opposed to before arriving to checkout.

This was originally discussed in P2: "wp-admin → checkout flow simplification": p1HpG7-onP-p2
**TLDR;-**
A Jetpack plugin user should be able to make a product purchase (initiated from My Jetpack, when disconnected) without having to manually connect the plugin first, but rather should be able to purchase the product first, then have the connection automatically initiated afterward.


See screencast video below to view the full "connect _**After**_ checkout" flow in action. ⤵️

## Proposed changes:

* Added a `getCalypsoOrigin()` helper in the connection package.
* Updated the My Jetpack product checkout flow (if site not connected/registered) to instead link to a connectAfterCheckout flow in Calypso as opposed to a connect "before" checkout flow.

### Other information:


## Jetpack product discussion



## Does this pull request change what data or activity we track or use?


No

## Testing instructions:






- Do one of these:
    - Checkout this PR and spin up a local development build of Jetpack, OR
    - Spin up a Jurassic.ninja site with the Jetpack Beta plugin running this PR.
- For minimal testing, skip to the section, "Basic Testing this PR only" below, otherwise follow the instructions to test the full checkout flow.

#### Testing the full checkout flow:
- Jetpack plugin activated, Not connected:
    - Go to My Jetpack:  `/wp-admin/admin.php?page=my-jetpack`
    - On the VaultPress Backup Card, click the "Purchase" Button.
    - Click "Get Jetpack VaultPress Backup" button.
    - Verify you are taken to: https://wordpress.com/checkout/jetpack/jetpack_backup_t1_yearly with at least these query args: `connect_after_checkout=true`, `admin_url=[your site admin url]`,  `from_site_slug=[your site slug]` and with Jetpack VaultPress Backup (10GB) in the cart. 
    - On the checkout page, use A81N credits, and select "Assign a payment method later".
    - Click "Complete checkout"
    - Verify purchase is completed and you are redirected to the Connection/Authorization screen.
    - Click "Authorize".
    - Verify connection completes successfully and you are redirected to Licensing Thank you auto-activation screen (`/checkout/jetpack/thank-you/licensing-auto-activate/jetpack_backup_t1_yearly?...`).
    -  See screencast video below to see the full "connect _After_ checkout" flow (these instructions) in action. ⤵️

#### Bonus points: - Test using PayPal

- Follow the instructions above, but instead of using A81N credits, use a test PayPal account (See: PCYsg-IA-p2  )

#### Basic Testing this PR only.
- Jetpack plugin activated, Not connected:
    - Go to:  `/wp-admin/admin.php?page=my-jetpack`
    - On the VaultPress Backup Card, click the "Purchase Button"
    - Click "Get Jetpack VaultPress Backup" button.
    - Verify you are taken to: https://wordpress.com/checkout/jetpack/jetpack_backup_t1_yearly with at least these query args: `connect_after_checkout=true`, `admin_url=[your site admin url]`,  `from_site_slug=[your site slug]`
 - Connect Jetpack:
     - Go to:  `/wp-admin/admin.php?page=my-jetpack`
    - On the VaultPress Backup Card, click the "Purchase Button"
    - Click "Get Jetpack VaultPress Backup" button.
    - Verify you are taken to: https://wordpress.com/checkout/[your site slug] (some query args will also be present) with Jetpack VaultPress Backup (10GB) in the cart. 

### Screencast: Connect _after_ checkout flow

https://github.com/Automattic/jetpack/assets/11078128/da0e540b-b722-4de3-927d-0952cdc3cbd5
